File: typescript/packages/tools/tests/subagent.test.ts
/**
 * SubagentTool tests — mirror `rust/crates/spore-core/src/tools/subagent.rs#tests`.
 * Uses a scripted Harness that returns a queue of `RunResult` values.
 */

import {
  emptyAggregateUsage,
  emptyBudgetSnapshot,
  emptySessionState,
  harnessTesting,
  newTask,
  SessionId,
  TaskId,
  toolRegistry,
  type Harness,
  type HarnessRunOptions,
  type HumanResponse,
  type PausedState,
  type RunResult,
  type StreamSink,
  type ToolCall,
} from "@spore/core";
import { describe, expect, it } from "vitest";

import { SubagentTool } from "../src/index.js";

const { AllowAllSandbox } = harnessTesting;
const { StandardToolRegistry, toolRegistryMock } = toolRegistry;
// Storage seam (#75): SubagentTool ignores ctx, but the signature requires one.
const ctx = toolRegistryMock.testCtx();

class ScriptedHarness implements Harness {
  constructor(private readonly results: RunResult[]) {}
  async run(_opts: HarnessRunOptions): Promise<RunResult> {
    const r = this.results.shift();
    if (!r) {
      return {
        kind: "failure",
        reason: { kind: "human_halted" },
        session_id: SessionId.of("s"),
        usage: emptyAggregateUsage(),
        turns: 0,
      };
    }
    return r;
  }
  async resume(
    _state: PausedState,
    _response: HumanResponse,
    _onStream?: StreamSink,
  ): Promise<RunResult> {
    return {
      kind: "failure",
      reason: { kind: "human_halted" },
      session_id: SessionId.of("s"),
      usage: emptyAggregateUsage(),
      turns: 0,
    };
  }
}

function callWith(input: unknown, id = "parent-call-1"): ToolCall {
  return { id, name: "subagent", input };
}

function buildSubagent(harness: Harness): SubagentTool {
  return SubagentTool.buildOrThrow({
    name: "subagent",
    description: "child",
    inputSchema: { type: "object" },
    timeoutMs: 5_000,
    contextSharing: { kind: "isolated" },
    harness,
    childRegistry: new StandardToolRegistry(),
  });
}

describe("SubagentTool", () => {
  it("maps child Success to ToolOutput.success", async () => {
    const harness = new ScriptedHarness([
      {
        kind: "success",
        output: "child done",
        session_id: SessionId.of("s"),
        usage: emptyAggregateUsage(),
        turns: 1,
      },
    ]);
    const sub = buildSubagent(harness);
    const r = await sub.execute(
      callWith({ instruction: "do it" }),
      new AllowAllSandbox(),
      ctx,
    );
    expect(r.kind).toBe("success");
    if (r.kind !== "success") throw new Error("unreachable");
    expect(r.content).toBe("child done");
  });

  it("maps child Failure to recoverable Error", async () => {
    const harness = new ScriptedHarness([
      {
        kind: "failure",
        reason: { kind: "human_halted" },
        session_id: SessionId.of("s"),
        usage: emptyAggregateUsage(),
        turns: 1,
      },
    ]);
    const sub = buildSubagent(harness);
    const r = await sub.execute(
      callWith({ instruction: "x" }),
      new AllowAllSandbox(),
      ctx,
    );
    expect(r.kind).toBe("error");
    if (r.kind !== "error") throw new Error("unreachable");
    expect(r.recoverable).toBe(true);
  });

  it("propagates WaitingForHuman with parent_tool_call_id", async () => {
    const sessionId = SessionId.of("s");
    const paused: PausedState = {
      session_id: sessionId,
      task_id: TaskId.of("t"),
      turn_number: 1,
      session_state: emptySessionState(),
      pending_tool_calls: [],
      approved_results: [],
      human_request: { kind: "clarification", question: "yes?" },
      task: newTask("x", sessionId, { kind: "re_act", max_iterations: 1 }),
      budget_used: emptyBudgetSnapshot(),
      child_state: null,
    };
    const harness = new ScriptedHarness([
      {
        kind: "waiting_for_human",
        state: paused,
        request: { kind: "clarification", question: "yes?" },
      },
    ]);
    const sub = buildSubagent(harness);
    const r = await sub.execute(
      callWith({ instruction: "x" }, "parent-call-1"),
      new AllowAllSandbox(),
      ctx,
    );
    expect(r.kind).toBe("waiting_for_human");
    if (r.kind !== "waiting_for_human") throw new Error("unreachable");
    expect(r.child_state.parent_tool_call_id).toBe("parent-call-1");
  });

  it("rejects construction when child registry has SubagentTool", () => {
    const childReg = new StandardToolRegistry();
    const err = childReg.register(
      new toolRegistryMock.SubagentMockTool("nested"),
      {
        name: "nested",
        description: "n",
        parameters: { type: "object" },
        annotations: {
          read_only: false,
          destructive: false,
          idempotent: false,
          open_world: false,
        },
      },
    );
    expect(err).toBeNull();
    const r = SubagentTool.build({
      name: "subagent",
      description: "child",
      inputSchema: { type: "object" },
      timeoutMs: 1_000,
      contextSharing: { kind: "isolated" },
      harness: new ScriptedHarness([]),
      childRegistry: childReg,
    });
    expect(r.ok).toBe(false);
    if (r.ok) throw new Error("unreachable");
    expect(r.error.kind).toBe("invalid_configuration");
  });

  it("missing instruction returns recoverable error", async () => {
    const sub = buildSubagent(new ScriptedHarness([]));
    const r = await sub.execute(callWith({}), new AllowAllSandbox(), ctx);
    expect(r.kind).toBe("error");
    if (r.kind !== "error") throw new Error("unreachable");
    expect(r.recoverable).toBe(true);
  });
});

File: go/spore-core/tools/subagent_test.go
package tools

import (
	"context"
	"encoding/json"
	"errors"
	"sync"
	"testing"
	"time"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// scriptedHarness yields a queue of pre-scripted RunResults. ResumeConsult logs
// the ConsultResponse it was resumed with and pops the next result (issue #114).
type scriptedHarness struct {
	mu        sync.Mutex
	results   []sporecore.RunResult
	resumeLog []sporecore.ConsultResponse
}

func (s *scriptedHarness) Run(_ context.Context, _ sporecore.HarnessRunOptions) sporecore.RunResult {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.results) == 0 {
		return sporecore.RunResult{
			Kind:   sporecore.RunFailure,
			Reason: sporecore.HaltReason{Kind: sporecore.HaltHumanHalted},
		}
	}
	r := s.results[0]
	s.results = s.results[1:]
	return r
}

func (s *scriptedHarness) Resume(_ context.Context, _ sporecore.PausedState, _ sporecore.HumanResponse, _ sporecore.StreamSink) sporecore.RunResult {
	return sporecore.RunResult{Kind: sporecore.RunFailure, Reason: sporecore.HaltReason{Kind: sporecore.HaltHumanHalted}}
}

func (s *scriptedHarness) ResumeConsult(_ context.Context, _ sporecore.PausedState, response sporecore.ConsultResponse, _ sporecore.StreamSink) sporecore.RunResult {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.resumeLog = append(s.resumeLog, response)
	if len(s.results) == 0 {
		return sporecore.RunResult{Kind: sporecore.RunSuccess, Output: "child done after consult"}
	}
	r := s.results[0]
	s.results = s.results[1:]
	return r
}

func newSubagent(t *testing.T, h sporecore.Harness, sharing ContextSharing) *SubagentTool {
	t.Helper()
	child := sporecore.NewStandardToolRegistry()
	s, err := NewSubagentTool("subagent", "child", json.RawMessage(`{"type":"object"}`),
		5*time.Second, sharing, h, child)
	if err != nil {
		t.Fatalf("NewSubagentTool: %v", err)
	}
	return s
}

func subagentCall(input any) sporecore.ToolCall {
	b, _ := json.Marshal(input)
	return sporecore.ToolCall{ID: "parent-call-1", Name: "subagent", Input: b}
}

func TestSubagentSuccessMaps(t *testing.T) {
	h := &scriptedHarness{results: []sporecore.RunResult{
		{Kind: sporecore.RunSuccess, Output: "child done"},
	}}
	s := newSubagent(t, h, Isolated{})
	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "do it"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputSuccess || r.Content != "child done" {
		t.Fatalf("%+v", r)
	}
}

func TestSubagentFailureMapsRecoverable(t *testing.T) {
	h := &scriptedHarness{results: []sporecore.RunResult{
		{Kind: sporecore.RunFailure, Reason: sporecore.HaltReason{Kind: sporecore.HaltHumanHalted}},
	}}
	s := newSubagent(t, h, Isolated{})
	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputError || !r.Recoverable {
		t.Fatalf("%+v", r)
	}
}

func TestSubagentWaitingForHumanPropagatesParentCallID(t *testing.T) {
	paused := &sporecore.PausedState{
		SessionID: "s",
		Task:      sporecore.NewTask("x", "s", sporecore.ReActStrategy(1)),
		HumanRequest: &sporecore.HumanRequest{
			Kind:     sporecore.HumanReqClarification,
			Question: "yes?",
		},
	}
	req := &sporecore.HumanRequest{Kind: sporecore.HumanReqClarification, Question: "yes?"}
	h := &scriptedHarness{results: []sporecore.RunResult{
		{Kind: sporecore.RunWaitingForHuman, State: paused, Request: req},
	}}
	s := newSubagent(t, h, Isolated{})
	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputWaitingForHuman {
		t.Fatalf("%+v", r)
	}
	if r.ChildState == nil || r.ChildState.ParentToolCallID != "parent-call-1" {
		t.Fatalf("child state wrong: %+v", r.ChildState)
	}
}

// subagentMockTool: a no-op tool that reports IsSubagentTool == true so
// HasSubagentTools() returns true on the child registry.
type subagentMockTool struct{}

func (subagentMockTool) Name() string                { return "nested_sub" }
func (subagentMockTool) IsSubagentTool() bool        { return true }
func (subagentMockTool) MayProduceLargeOutput() bool { return false }
func (subagentMockTool) Execute(context.Context, sporecore.ToolCall, sporecore.SandboxProvider, *sporecore.ToolContext) sporecore.ToolOutput {
	return sporecore.ToolOutput{Kind: sporecore.ToolOutputSuccess}
}

func TestSubagentConstructionRejectsNestedChild(t *testing.T) {
	child := sporecore.NewStandardToolRegistry()
	schema := sporecore.RegistryToolSchema{
		Name:        "nested_sub",
		Description: "n",
		Parameters:  json.RawMessage(`{"type":"object"}`),
	}
	if err := child.Register(subagentMockTool{}, schema); err != nil {
		t.Fatalf("register: %v", err)
	}
	h := &scriptedHarness{}
	_, err := NewSubagentTool("subagent", "child", json.RawMessage(`{"type":"object"}`),
		time.Second, Isolated{}, h, child)
	if err == nil {
		t.Fatalf("expected build error")
	}
	if !errors.Is(err, ErrInvalidConfiguration) {
		t.Fatalf("expected ErrInvalidConfiguration, got %v", err)
	}
	var be *BuildError
	if !errors.As(err, &be) {
		t.Fatalf("expected *BuildError, got %T", err)
	}
}

func TestSubagentMissingInstructionRecoverable(t *testing.T) {
	h := &scriptedHarness{}
	s := newSubagent(t, h, Isolated{})
	r := s.Execute(context.Background(), subagentCall(map[string]any{}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputError || !r.Recoverable {
		t.Fatalf("%+v", r)
	}
}

func TestContextSharingJSONRoundtrip(t *testing.T) {
	for _, c := range []ContextSharing{
		Isolated{},
		SharedSession{SessionID: "s1"},
		SummaryHandoff{Summary: "did stuff"},
	} {
		b, err := MarshalContextSharing(c)
		if err != nil {
			t.Fatalf("marshal: %v", err)
		}
		back, err := UnmarshalContextSharing(b)
		if err != nil {
			t.Fatalf("unmarshal: %v", err)
		}
		if back.Kind() != c.Kind() {
			t.Fatalf("kind mismatch: %s vs %s", back.Kind(), c.Kind())
		}
	}
}
